### PR TITLE
Create Audio Feature in SDK

### PR DESCRIPTION
**Pull request by nickyfantasy**
* Add apis to record audio in SDK
* Add corresponding apis in pybind, storage.py, sdk.h
* Implement reservoir sampling when collecting audio samples

### CODE REVIEW

**Comment by jetfuel** [comment]
@nickyfantasy Please run pre-commit to format the code. The Travis CI will check the format.

File: visualdl/logic/sdk.cc [hunk 1]
```diff
@@ -352,6 +347,112 @@ std::string TextReader::caption() const {
...
+  auto idx = IsSampleTaken();
+  if (idx >= 0) {
+    SetSample(idx, sample_rate, data);
+  }
+}
+
+void Audio::SetSample(int index,
+                      int sample_rate,
+                      const std::vector<value_t>& data) {
+  CHECK_GT(sample_rate, 0) << "sample rate should be something like 6000, 8000 or 44100";
```
Comment (daming-lu) [inline visualdl/logic/sdk.cc]: should be a positive number?

File: visualdl/logic/sdk.cc [hunk 2]
```diff
@@ -352,6 +347,112 @@ std::string TextReader::caption() const {
...
+  CHECK_LT(index, num_samples_);
```
Comment (daming-lu) [inline visualdl/logic/sdk.cc]: We can add error messages for these 2 as well.

File: visualdl/logic/sdk.h [hunk 3]
```diff
@@ -326,6 +326,123 @@ struct TextReader {
...
+/*
+* Audio reader.
+*/
+struct AudioReader {
+    using value_t = typename Audio::value_t;
+
+    struct AudioRecord {
+        int step_id;
+        int sample_rate;
+        std::vector<int> data;
```
Comment (daming-lu) [inline visualdl/logic/sdk.h]: So here the data is 'int' instead of 'float'?

File: visualdl/logic/sdk.cc [hunk 4]
```diff
@@ -352,6 +347,112 @@ std::string TextReader::caption() const {
...
+AudioReader::AudioRecord AudioReader::record(int offset, int index) {
+  AudioRecord res;
+  auto record = reader_.record(offset);
+  auto entry = record.data(index);
+  auto filename = entry.GetRaw();
+  CHECK(!g_log_dir.empty())
+          << "g_log_dir should be set in LogReader construction";
+  BinaryRecordReader brcd(GenBinaryRecordDir(g_log_dir), filename);
+
+  std::transform(brcd.data.begin(),
```
Comment (daming-lu) [inline visualdl/logic/sdk.cc]: is brcd.data the same as res.data?

File: visualdl/logic/sdk.h [hunk 5]
```diff
@@ -326,6 +326,123 @@ struct TextReader {
...
+     */
+    void StartSampling();
+    /*
+     * End a sampling period, it will clear all states for reservior sampling.
+     */
+    void FinishSampling();
+
+    /*
+     * A combined interface for IsSampleTaken and SetSample, simpler but might be
+     * low effience.
```
Comment (daming-lu) [inline visualdl/logic/sdk.h]: efficiency

File: visualdl/python/storage.py [hunk 6]
```diff
@@ -119,6 +119,16 @@ def text(self, tag):
         check_tag_name_valid(tag)
         return self.reader.get_text(tag)
 
+    def audio(self, tag):
+        """
+        Get a audio reader with tag
```
Comment (daming-lu) [inline visualdl/python/storage.py]: Get 'an' audio

File: visualdl/logic/pybind.cc [hunk 7]
```diff
@@ -74,14 +74,17 @@ PYBIND11_MODULE(core, m) {
       #undef READER_ADD_HISTOGRAM
 
       // clang-format on
-      .def("get_image",
-           [](vs::LogReader& self, const std::string& tag) {
-             auto tablet = self.tablet(tag);
-             return vs::components::ImageReader(self.mode(), tablet);
-           })
+      .def("get_image", [](vs::LogReader& self, const std::string& tag) {
```
Comment (Superjomn) [inline visualdl/logic/pybind.cc]: these seem just some code style formatting, and we each other have a different clang-format config that results in some diff, maybe we need a unified version of config that makes we have the same code style?

In paddle, the clang-format 3.8 and google c++ style is used, different config and version may lead to some diff.

We can reference [paddle's .clang-format configuration](https://github.com/PaddlePaddle/Paddle/blob/develop/.clang-format)

File: visualdl/logic/sdk.cc [hunk 8]
```diff
@@ -352,6 +347,112 @@ std::string TextReader::caption() const {
...
+
+void Audio::SetSample(int index,
+                      int sample_rate,
+                      const std::vector<value_t>& data) {
+  CHECK_GT(sample_rate, 0) << "sample rate should be something like 6000, 8000 or 44100";
+  CHECK_LT(index, num_samples_);
+  CHECK_LE(index, num_records_);
+
+    //convert float vector to char vector
+  std::vector<char> data_str(data.size());
```
Comment (Superjomn) [inline visualdl/logic/sdk.cc]: it seems that data_str can directly be a `string` and no need to tranform from `vector` to `string` again.

```c++
std::string data_str(data.size());
...
BinaryRecord brcd(xxdir, std::move(data_str));
```

File: visualdl/logic/sdk.h (hunk 3 above)
Comment (Superjomn) [inline visualdl/logic/sdk.h]: To meet the audio value interval, is `short` or `unsigned char` or `char` is enough? 

https://cn.mathworks.com/help/matlab/ref/audiorecorder.getaudiodata.html?s_tid=gn_loc_drop

here just `int16`, `int8`, `uint8` are used, not `int32`.

Just a suggestion, not that important, `int` works good.

File: visualdl/logic/sdk.cc (hunk 1 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.cc]: yes

File: visualdl/logic/sdk.cc (hunk 2 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.cc]: sure

File: visualdl/logic/sdk.h (hunk 3 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.h]: yes, when we write data in, we convert float to string, when we read the data out, we convert binary to int

File: visualdl/logic/sdk.cc (hunk 4 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.cc]: brcd.data is the data in string format when we saved in file, when we read the data we convert to integer that becomes res.data

File: visualdl/logic/pybind.cc (hunk 7 above)
Comment (nickyfantasy) [reply visualdl/logic/pybind.cc]: yes, I just updated clang-format

File: visualdl/logic/sdk.cc (hunk 8 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.cc]: ok, I end up just use ```std::string(data.begin(),data.end())``` to directly convert the data vector to string

File: visualdl/logic/sdk.h (hunk 3 above)
Comment (nickyfantasy) [reply visualdl/logic/sdk.h]: you are right, when we were doing speech recognition app before, we just use byte / int8

File: visualdl/logic/pybind.cc [hunk 9]
```diff
@@ -219,6 +233,61 @@ PYBIND11_MODULE(core, m) {
       .def("total_records", &cp::TextReader::total_records)
       .def("size", &cp::TextReader::size);
 
+  py::class_<cp::Audio>(m, "AudioWriter", R"pbdoc(
```
Comment (jetfuel) [inline visualdl/logic/pybind.cc:236]: Will it be weird to have documentations published on the website but not the code is not in the release pip? I am not sure what's the best approach here.

File: visualdl/logic/pybind.cc [hunk 10]
```diff
@@ -219,6 +233,61 @@ PYBIND11_MODULE(core, m) {
...
+            A combined interface for is_sample_taken and set_sample, simpler but is less efficient.
+
+            :param sample_rate: Sample rate of audio
+            :type sample_rate: integer
+            :param audio_data: Flatten audio data
+            :type audio_data: list
+                  )pbdoc");
+
+  py::class_<cp::AudioReader::AudioRecord>(m, "AudioRecord")
+      // TODO(ChunweiYan) make these copyless.
```
Comment (jetfuel) [inline visualdl/logic/pybind.cc]: Either remove the TODO or update it to yours

File: visualdl/logic/sdk.cc [hunk 11]
```diff
@@ -352,6 +347,111 @@ std::string TextReader::caption() const {
...
+  step_.SetTimeStamp(time);
+
+  // resize record
+  for (int i = 0; i < num_samples_; i++) {
+    step_.AddData();
+  }
+  num_records_ = 0;
+}
+
+int Audio::IsSampleTaken() {
```
Comment (jetfuel) [inline visualdl/logic/sdk.cc]: Minor stuff, the function name is implying that the function will return a BOOL, but the function returns an index. 
Maybe rename the function to `NextRandSampleIndex` or provide a comment here to explain the logic.

Comment (nickyfantasy) [reply visualdl/logic/sdk.cc]: true

File: visualdl/logic/pybind.cc (hunk 10 above)
Comment (nickyfantasy) [reply visualdl/logic/pybind.cc]: ya

**Review by daming-lu — DISMISSED** [review]

**Review by daming-lu — APPROVED** [review]

**Pull request merged**